File: packages/i18nify-go/modules/datetime/get_primary_timezone.go
package datetime

import (
	"fmt"
	"strings"
)

// GetPrimaryTimezone returns the capital city's IANA timezone for a country code.
func GetPrimaryTimezone(countryCode string) (string, error) {
	code := strings.ToUpper(strings.TrimSpace(countryCode))
	if code == "" {
		return "", fmt.Errorf("getPrimaryTimezone: country code must not be empty")
	}

	metadata, err := getCountryMetadataData()
	if err != nil {
		return "", fmt.Errorf("getPrimaryTimezone: %w", err)
	}

	countryMeta, exists := metadata.GetMetadataInformation()[code]
	if !exists || countryMeta == nil {
		return "", fmt.Errorf("getPrimaryTimezone: country code %q not found", code)
	}

	timezone := countryMeta.GetTimezoneOfCapital()
	if timezone == "" {
		return "", fmt.Errorf("getPrimaryTimezone: no primary timezone for country code %q", code)
	}

	return timezone, nil
}

File: packages/i18nify-js/src/modules/dateTime/getPrimaryTimezone.ts
import { withErrorBoundary } from '../../common/errorBoundary';
import { I18NIFY_DATA_SOURCE } from '../shared';

const getPrimaryTimezone = (countryCode: string): Promise<string> => {
  if (!countryCode || !countryCode.trim()) {
    return Promise.reject(
      new Error('getPrimaryTimezone: country code must not be empty.'),
    );
  }

  const normalizedCode = countryCode.trim().toUpperCase();

  return fetch(`${I18NIFY_DATA_SOURCE}/country/metadata/data.json`)
    .then((response) => {
      if (!response.ok) {
        throw new Error(`HTTP ${response.status}: ${response.statusText}`);
      }

      return response.json();
    })
    .then((response) => {
      const countryMeta = response.metadata_information?.[normalizedCode];

      if (!countryMeta) {
        throw new Error(
          `Country code "${normalizedCode}" not found in metadata.`,
        );
      }

      const timezone: string = countryMeta.timezone_of_capital ?? '';

      if (!timezone) {
        throw new Error(
          `No primary timezone found for country code "${normalizedCode}".`,
        );
      }

      return timezone;
    })
    .catch((error) => {
      throw new Error(
        `An error occurred while fetching primary timezone for country "${countryCode}". The error details are: ${error.message}.`,
      );
    });
};

export default withErrorBoundary<typeof getPrimaryTimezone>(getPrimaryTimezone);
